nlpia_bot/clibot.py

```diff
@@ -187,19 +192,20 @@ def reply(self, statement=''):
                     log.error(str(e))
             bot_replies = normalize_replies(bot_replies)
             replies.extend(bot_replies)
+
+        # Weighted random selection of reply from those with top n confidence scores 
         if len(replies):
             log.info(f'Found {len(replies)} suitable replies, limiting to {self.num_top_replies}...')
             replies = self.quality_score.update_replies(replies, statement)
             replies = sorted(replies, reverse=True)[:self.num_top_replies]
-            cumsum = 0
-            cdf = list()
-            for reply in replies:
-                cumsum += reply[0]
-                cdf.append(cumsum)
-            roll = np.random.rand() * cumsum
-            for i, threshold in enumerate(cdf):
+            
+            confidences, strings = list(zip(*replies))
```

Comment:
I love it! Generally prefer `texts` over `strings` as a variable name for natural language text lists/arrays.